src/dynamics/mpas/driver/dyn_mpas_subdriver.F90

```diff
@@ -0,0 +1,482 @@
+    !
+    ! This module manages the life cycle (i.e., initialization, running, and
+    ! finalization) of MPAS as a dynamical core within CAM-SIMA.
+    !
+    !-------------------------------------------------------------------------------
+
+    use, intrinsic :: iso_fortran_env, only: output_unit
+
+    ! Modules from external libraries.
+    use mpi, only: mpi_comm_null, mpi_comm_rank, mpi_success
+    use pio, only: iosystem_desc_t, pio_iosystem_is_active
+
+    ! Modules from MPAS.
+    use atm_core_interface, only: atm_setup_core, atm_setup_domain
+    use mpas_derived_types, only: core_type, domain_type
+    use mpas_domain_routines, only: mpas_allocate_domain
+    use mpas_framework, only: mpas_framework_init_phase1, mpas_framework_init_phase2
+    use mpas_kind_types, only: strkind
+    use mpas_pool_routines, only: mpas_pool_get_config
+
+    implicit none
+
+    private
+    public :: mpas_dynamical_core_type
+
+    abstract interface
+        ! This interface is compatible with `endrun` from CAM-SIMA.
+        subroutine model_error_if(message, file, line)
+            character(*),           intent(in) :: message
+            character(*), optional, intent(in) :: file
+            integer,      optional, intent(in) :: line
+        end subroutine model_error_if
+    end interface
+
+    !> The "class" of MPAS dynamical core.
+    !> Important data structures like states of MPAS dynamical core are encapsulated inside this derived type to prevent misuse.
+    !> Type-bound procedures provide well-defined APIs for CAM-SIMA to interact with MPAS dynamical core.
+    type :: mpas_dynamical_core_type
+        private
+
+        integer :: log_unit = output_unit
+        integer :: mpi_comm = mpi_comm_null
+        integer :: mpi_rank = 0
+        logical :: mpi_rank_master = .false.
+
+        ! Actual implementation is supplied at runtime.
+        procedure(model_error_if), nopass, pointer :: model_error => null()
+
+        type(core_type), pointer :: corelist => null()
+        type(domain_type), pointer :: domain_ptr => null()
+    contains
+        private
+
+        procedure, pass, public :: debug_print => dyn_mpas_debug_print
+        procedure, pass, public :: init_phase1 => dyn_mpas_init_phase1
+        procedure, pass, public :: read_namelist => dyn_mpas_read_namelist
+        procedure, pass, public :: init_phase2 => dyn_mpas_init_phase2
+    end type mpas_dynamical_core_type
+contains
+    !> Print a debug message with optionally the value(s) of a variable.
+    !> If `printer` is not supplied, the MPI master rank will print. Otherwise, the designated MPI rank will print instead.
+    !> (KCW, 2024-02-03)
+    subroutine dyn_mpas_debug_print(self, message, variable, printer)
+        class(mpas_dynamical_core_type), intent(in) :: self
+        character(*), intent(in) :: message
+        class(*), optional, intent(in) :: variable(:)
+        integer, optional, intent(in) :: printer
+
+#ifndef NDEBUG
```

Comment:
Perhaps it's a matter of personal preference (i.e., I'm not insisting on changes here), but would it be clearer to avoid a double-negative and use
```
#ifdef DEBUG
```
instead?

Reply:
I do agree that the double-negative is not quite straightforward. However, the rationale to use the `NDEBUG` macro is that it is part of the C/C++ standard [[1]](https://en.cppreference.com/w/c/error/assert)[[2]](https://en.cppreference.com/w/cpp/error/assert).

Reply:
Since this is Fortran code, and since the `NDEBUG` macro in C/C++ seems to be specifically intended for controlling `assert` (as far as I'm aware; I may be wrong), I still think it might be worth reconsidering. There are arguments against using `NDEBUG` for anything other than controlling assertions; for example, perhaps we would like to keep assertions enabled (in C/C++ code) but to not have a large volume of log output.

Ultimately, there may be a better time and place to have this discussion -- how to control debugging log output, perhaps with some finer granularity than "all or nothing".

Reply:
I pushed https://github.com/ESCOMP/CAM-SIMA/pull/248/commits/361d187517efdf343d14a4f6dd2b133d96f4ef60, which should address this.